internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+package gcs
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/guid"
+	"github.com/sirupsen/logrus"
+)
+
+type requestMessage interface {
+	SetActivityID(id guid.GUID)
+}
+
+type responseMessage interface {
+	Err() error
+}
+
+// rpc represents an outstanding rpc request to the guest
+type rpc struct {
+	proc rpcProc
+	id   int64
+	req  requestMessage
+	resp responseMessage
+	err  error
+	ch   chan struct{}
+}
+
+// bridge represents a communcations bridge with the guest. It handles the
+// transport layer but (mostly) does not parse or construct the message payload.
+type bridge struct {
+	Timeout time.Duration
+	mu      sync.Mutex
+	nextID  int64
+	rpcs    map[int64]*rpc
+	conn    io.ReadWriteCloser
+	rpcCh   chan *rpc
+	notify  notifyFunc
+	closed  bool
+	log     *logrus.Entry
+}
+
+var (
+	errBridgeClosed = errors.New("bridge closed")
+
+	// Any non-async RPC must respond in this time or the bridge
+	// connection will be terminated.
+	bridgeFailureTimeout = time.Minute * 5
+)
+
+type notifyFunc func(*containerNotification) error
+
+// newBridge returns a bridge on the given connection. It calls notify when a
+// notification message arrives from the guest. It logs transport errors and
+// traces using the given logrus entry.
+func newBridge(conn io.ReadWriteCloser, notify notifyFunc, log *logrus.Entry) *bridge {
+	return &bridge{
+		conn:    conn,
+		rpcs:    make(map[int64]*rpc),
+		rpcCh:   make(chan *rpc),
+		notify:  notify,
+		log:     log,
+		Timeout: bridgeFailureTimeout,
+	}
+}
+
+// Start begins the bridge send and receive goroutines.
+func (brdg *bridge) Start() {
+	go brdg.recvLoopRoutine()
+	go brdg.sendLoopRoutine()
+}
+
+// Kill terminates the bridge, closing the connection and causing all new and
+// existing RPCs to fail.
+func (brdg *bridge) Kill() {
+	brdg.mu.Lock()
+	if brdg.closed {
+		brdg.mu.Unlock()
+		return
+	}
+	brdg.closed = true
+	brdg.mu.Unlock()
+	brdg.log.Debug("bridge terminating")
+	brdg.conn.Close()
+}
+
+// Close closes the bridge. It is undefined to call RPC or AsyncRPC after
+// calling Close.
+func (brdg *bridge) Close() error {
+	brdg.Kill()
+	close(brdg.rpcCh)
+	return nil
+}
+
+// AsyncRPC sends an RPC request to the guest but does not wait for a response.
+// If the message cannot be sent before the context is done, then an error is
+// returned.
+func (brdg *bridge) AsyncRPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage) (*rpc, error) {
+	call := &rpc{
+		ch:   make(chan struct{}),
+		proc: proc,
+		req:  req,
+		resp: resp,
+	}
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	// Send the request.
+	select {
+	case brdg.rpcCh <- call:
+		return call, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (call *rpc) complete(err error) {
```

Comment:
I know its overkill but can we move `rpc` to `rpc.go` to keep it inline. Will make it easier to have a `_test.go` file for it as well.

Reply:
Hmm. I think rpc is inexorably linked to bridge, so separating it doesn't make a lot of sense. The RPC tests are part of the bridge tests.

This isn't C#, you know?

Reply:
I asked you to put them in a different file in the same package. Not to make a new dynamically linked dll with an `InternalsVisibleTo` attribute...